No updates for three days

Since 2022-07-20 we get notifications that GND updates have failed. I will have to investigate whether this is a problem on our side or on at the data provider (DNB). 

@thoffma Do you know anything about problems with the OAI-PMH updates for GND RDF?

Did manual update like noted in the README. After some successful OAI-PMH harvesting with several resumptionTokens, it exited with: 

```
[...]
resumptionToken: 
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at org.apache.xerces.dom.DeferredDocumentImpl.getNodeObject(Unknown Source)
	at org.apache.xerces.dom.DeferredDocumentImpl.synchronizeChildren(Unknown Source)
	at org.apache.xerces.dom.DeferredElementNSImpl.synchronizeChildren(Unknown Source)
	at org.apache.xerces.dom.ParentNode.hasChildNodes(Unknown Source)
	at org.apache.xerces.dom.DeepNodeListImpl.nextMatchingElementAfter(Unknown Source)
	at org.apache.xerces.dom.DeepNodeListImpl.item(Unknown Source)
	at org.joox.Impl.find(Impl.java:512)
	at org.joox.Impl.find(Impl.java:477)
	at org.joox.Impl.find(Impl.java:79)
	at apps.ConvertUpdates.process(ConvertUpdates.java:112)
	at apps.ConvertUpdates.getUpdates(ConvertUpdates.java:89)
	at apps.ConvertUpdates.getUpdatesAndConvert(ConvertUpdates.java:69)
[...]
[ERROR] [07/28/2022 16:59:54.784] [sbt-web-scheduler-1] [ActorSystem(sbt-web)] exception on LARS’ timer thread
java.lang.OutOfMemoryError: GC overhead limit exceeded
```
After some increasing steps of java heap space I ended up with `java -Xmx16G`. Nevertheless the `GC overhead limit exceeded`. However, the daily cronjob of updates at 1:00 am ended successfully. Also @acka47 could confirm that new data is available.
So gonna close this.

--- a/README.textile
+++ b/README.textile
@@ -102,9 +102,9 @@ h4. Get and convert the updates
 
 Updates are pulled via "the DNB OAI-PMH interface":http://www.dnb.de/DE/Service/DigitaleDienste/OAI/oai_node.html.
 
-Pass one argument: get updates since a given date:
+Pass one or two arguments: get updates since (and optionally until) a given date:
 
-@sbt "runMain apps.ConvertUpdates 2020-06-22"@
+@sbt "runMain apps.ConvertUpdates 2022-06-22 2022-06-23"@
 
 The date of the most recent update is stored in 'GND-lastSuccessfulUpdate.txt' (can be changed in the config).
 

--- a/app/apps/ConvertUpdates.java
+++ b/app/apps/ConvertUpdates.java
@@ -17,13 +17,23 @@ import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
 import java.time.temporal.ChronoUnit;
+import java.util.Iterator;
 
+import javax.xml.namespace.QName;
 import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.stream.FactoryConfigurationError;
+import javax.xml.stream.XMLEventFactory;
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.XMLEventWriter;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.events.EndElement;
+import javax.xml.stream.events.StartElement;
+import javax.xml.stream.events.XMLEvent;
 import javax.xml.transform.TransformerException;
 
 import org.apache.commons.lang3.tuple.Pair;
-import org.joox.JOOX;
-import org.joox.Match;
 import org.xml.sax.SAXException;
 
 import ORG.oclc.oai.harvester2.app.RawWrite;
@@ -110,12 +120,51 @@ public class ConvertUpdates {
 	}
 
 	public static void process(final String baseUrl, String from, String until, File result)
-			throws NoSuchFieldException, IOException, ParserConfigurationException, SAXException, TransformerException {
+			throws NoSuchFieldException, IOException, ParserConfigurationException, SAXException, TransformerException,
+			XMLStreamException {
 		ByteArrayOutputStream stream = new ByteArrayOutputStream();
 		System.out.printf("Calling OAI-PMH at %s from %s until %s\n", baseUrl, from, until);
 		RawWrite.run(baseUrl, from, until, "RDFxml", "authorities", stream);
-		Match m = JOOX.$(new InputStreamReader(new ByteArrayInputStream(stream.toByteArray()), StandardCharsets.UTF_8));
-		m.find("Description").write(new FileWriter(result, true));
+		writeRdfDescriptions(result, stream);
+	}
+
+	static void writeRdfDescriptions(File result, ByteArrayOutputStream stream)
+			throws FactoryConfigurationError, XMLStreamException, IOException {
+		try (FileWriter fileWriter = new FileWriter(result, true)) {
+			String rdfTag = "RDF";
+			String entityTag = "Description";
+			XMLOutputFactory outputFactory = XMLOutputFactory.newInstance();
+			XMLEventFactory eventFactory = XMLEventFactory.newInstance();
+			XMLEventWriter eventWriter = null;
+			InputStreamReader inputStreamReader = new InputStreamReader(new ByteArrayInputStream(stream.toByteArray()),
+					StandardCharsets.UTF_8);
+			XMLEventReader eventReader = XMLInputFactory.newInstance().createXMLEventReader(inputStreamReader);
+			Iterator<?> namespaces = null;
+			while (eventReader.hasNext()) {
+				XMLEvent nextEvent = eventReader.nextEvent();
+				if (nextEvent.isStartElement()) {
+					StartElement startElement = nextEvent.asStartElement();
+					if (startElement.getName().getLocalPart().equals(rdfTag)) {
+						namespaces = startElement.getNamespaces();
+						eventWriter = outputFactory.createXMLEventWriter(fileWriter);
+						continue;
+					} else if (startElement.getName().getLocalPart().equals(entityTag)) {
+						QName descriptionName = new QName(startElement.getName().getNamespaceURI(),
+								startElement.getName().getLocalPart(), startElement.getName().getPrefix());
+						StartElement descriptionWithNamespaces = eventFactory.createStartElement(descriptionName,
+								startElement.getAttributes(), namespaces);
+						nextEvent = descriptionWithNamespaces;
+					}
+				} else if (nextEvent.isEndElement()
+						&& ((EndElement) nextEvent).getName().getLocalPart().equals(rdfTag)) {
+					eventWriter.close();
+					eventWriter = null;
+				}
+				if (eventWriter != null) {
+					eventWriter.add(nextEvent);
+				}
+			}
+		}
 	}
 
 	private static void writeLastSuccessfulUpdate(String until) {

--- a/app/apps/ConvertUpdates.java
+++ b/app/apps/ConvertUpdates.java
@@ -37,8 +37,9 @@ public class ConvertUpdates {
 			+ "This may or may not be a problem on the side of the data provider.";
 
 	public static void main(String[] args) throws IOException, InterruptedException {
-		if (args.length == 1) {
-			Pair<String, String> startAndEnd = getUpdatesAndConvert(args[0]);
+		if (args.length == 1 || args.length == 2) {
+			String endOfUpdates = args.length == 2 ? args[1] : null;
+			Pair<String, String> startAndEnd = getUpdatesAndConvert(args[0], endOfUpdates);
 			File dataUpdate = new File(config("data.updates.data"));
 			short tried = 0;
 			while (dataUpdate.length() == 0 ) {
@@ -49,7 +50,7 @@ public class ConvertUpdates {
 					break;
 				System.err.println("Going to retry in " + WAIT_PER_RETRY / 1000 / 60 + " min");
 				Thread.sleep(WAIT_PER_RETRY);
-				startAndEnd = getUpdatesAndConvert(args[0]);
+				startAndEnd = getUpdatesAndConvert(args[0], endOfUpdates);
 			}
 			if (dataUpdate.length() == 0) {
 				System.err.println("Tried " + tried
@@ -61,23 +62,27 @@ public class ConvertUpdates {
 			}
 			backup(dataUpdate, startAndEnd.getLeft(), startAndEnd.getRight());
 		} else {
-			System.err.println("Argument missing to get updates since a given date in ISO format, e.g. 2019-06-13");
+			System.err.println(
+					"Argument missing to get updates since (and optionally until) a given date in ISO format, e.g. 2019-06-13");
 		}
 	}
 
-	private static Pair<String, String> getUpdatesAndConvert(final String startOfUpdates) throws IOException {
-		Pair<String, String> startAndEnd = getUpdates(startOfUpdates);
+	private static Pair<String, String> getUpdatesAndConvert(final String startOfUpdates, final String endOfUpdates) throws IOException {
+		Pair<String, String> startAndEnd = getUpdates(startOfUpdates, endOfUpdates);
 		backup(new File(config("data.updates.rdf")), startAndEnd.getLeft(), startAndEnd.getRight());
 		ConvertBaseline.main(new String[] { config("data.updates.rdf"), config("data.updates.data"),
 				config("index.delete.updates") });
 		return startAndEnd;
 	}
 	
-	private static Pair<String, String> getUpdates(String startOfUpdates) {
+	private static Pair<String, String> getUpdates(String startOfUpdates, String endOfUpdates) {
 		int intervalSize = Convert.CONFIG.getInt("data.updates.interval");
 		String start = startOfUpdates;
-		String end = addDays(start, intervalSize);
-		int intervals = calculateIntervals(startOfUpdates, intervalSize);
+		String givenEndOrToday = endOfUpdates != null ? endOfUpdates
+				: LocalDateTime.now().format(DateTimeFormatter.ISO_DATE);
+		String end = addDays(start, givenEndOrToday, intervalSize);
+		int intervals = calculateIntervals(startOfUpdates, givenEndOrToday,
+				intervalSize);
 		File file = new File(config("data.updates.rdf"));
 		try (FileWriter writer = new FileWriter(file, false)) {
 			writer.write("<RDF>");
@@ -90,11 +95,11 @@ public class ConvertUpdates {
 			} catch (Throwable t) {
 				t.printStackTrace();
 			}
-			start = addDays(start, intervalSize);
+			start = addDays(start, givenEndOrToday, intervalSize);
 			if (i == intervals - 2)
-				end = getToday();
+				end = givenEndOrToday;
 			else
-				end = addDays(end, intervalSize);
+				end = addDays(end, givenEndOrToday, intervalSize);
 		}
 		try (FileWriter writer = new FileWriter(file, true)) {
 			writer.write("</RDF>");
@@ -107,6 +112,7 @@ public class ConvertUpdates {
 	public static void process(final String baseUrl, String from, String until, File result)
 			throws NoSuchFieldException, IOException, ParserConfigurationException, SAXException, TransformerException {
 		ByteArrayOutputStream stream = new ByteArrayOutputStream();
+		System.out.printf("Calling OAI-PMH at %s from %s until %s\n", baseUrl, from, until);
 		RawWrite.run(baseUrl, from, until, "RDFxml", "authorities", stream);
 		Match m = JOOX.$(new InputStreamReader(new ByteArrayInputStream(stream.toByteArray()), StandardCharsets.UTF_8));
 		m.find("Description").write(new FileWriter(result, true));
@@ -129,19 +135,14 @@ public class ConvertUpdates {
 		Files.copy(source.toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING);
 	}
 
-	private static String addDays(String start, int intervalSize) {
+	private static String addDays(String start, String end, int intervalSize) {
 		LocalDate endDate = LocalDate.parse(start).plusDays(intervalSize);
-		return LocalDate.parse(getToday()).isBefore(endDate) ? getToday() : endDate.format(DateTimeFormatter.ISO_DATE);
+		return LocalDate.parse(end).isBefore(endDate) ? end : endDate.format(DateTimeFormatter.ISO_DATE);
 	}
 
-	private static String getToday() {
-		return LocalDateTime.now().format(DateTimeFormatter.ISO_DATE);
-	}
-
-	private static int calculateIntervals(String startOfUpdates, int intervalSize) {
-		String end = getToday();
+	private static int calculateIntervals(String startOfUpdates, String endOfUpdates, int intervalSize) {
 		final LocalDate startDate = LocalDate.parse(startOfUpdates);
-		final LocalDate endDate = LocalDate.parse(end);
+		final LocalDate endDate = LocalDate.parse(endOfUpdates);
 		long timeSpan = startDate.until(endDate, ChronoUnit.DAYS);
 		return ((int) timeSpan / intervalSize) + 1;
 	}

--- a/test/apps/ConvertUpdatesTest.java
+++ b/test/apps/ConvertUpdatesTest.java
@@ -1,0 +1,31 @@
+package apps;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+import javax.xml.stream.FactoryConfigurationError;
+import javax.xml.stream.XMLStreamException;
+
+import org.junit.Test;
+
+public class ConvertUpdatesTest {
+
+	@Test
+	public void testXmlParsing() throws IOException, FactoryConfigurationError, XMLStreamException {
+		ByteArrayOutputStream stream = new ByteArrayOutputStream();
+		String xml = "<x><m><l>"
+				+ "<RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\"><rdf:Description><r><d><f1/></d></r><more1/></rdf:Description></RDF>"
+				+ "<RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\"><rdf:Description><r><d><f2/></d></r><more2/></rdf:Description></RDF>"
+				+ "</l></m></x>";
+		stream.write(xml.getBytes());
+		File result = File.createTempFile("parsed", ".rdf");
+		result.deleteOnExit();
+		ConvertUpdates.writeRdfDescriptions(result, stream);
+		System.out.println("Wrote to " + result + ": ");
+		Files.readAllLines(Paths.get(result.toURI())).forEach(System.out::println);
+	}
+
+}

--- a/test/apps/DataWorkflowTests.java
+++ b/test/apps/DataWorkflowTests.java
@@ -7,7 +7,7 @@ import org.junit.runners.Suite.SuiteClasses;
 import modules.IndexQueryTest;
 
 @RunWith(Suite.class)
-@SuiteClasses({ ConvertTest.class, IndexAppTest.class, IndexQueryTest.class })
+@SuiteClasses({ ConvertTest.class, ConvertUpdatesTest.class, IndexAppTest.class, IndexQueryTest.class })
 public class DataWorkflowTests {
 
 }

